### ISSUE TITLE
FR: Show an alert message if "crop and rotate" module is focused and "crop" ist already switched on

### ISSUE THREAD

**Issue by domosbg** [darktable-org/darktable#9846]
Started as a bug request, this issue was changed to a feature request...

**Did you buy darktable from an application store ?**  -> NO

**Describe the bug/issue**

**To Reproduce**
_Please provide detailed steps to reproduce the behaviour, for example:_
1. Open picture in Darkroom
![grafik](https://user-images.githubusercontent.com/88573684/130294015-59398288-d174-47bb-b64d-db7de19a608c.png)

2. Crop picture by "crop" module
![grafik](https://user-images.githubusercontent.com/88573684/130294095-63fc7eb3-11af-4b34-8ba3-21c5aee3e462.png)

3. Click on module "crop and rotate"
4. The complete picture is shown, not the cropped one
![grafik](https://user-images.githubusercontent.com/88573684/130294172-a37639dd-95f3-4fcd-9a18-ffe92cc4991b.png)

If you crop it again, it results in a double cropped picture, but I couldn't find the system how the second crop is done on the picture.

**Expected behavior**
If crop and rotate is activated, the cropped pictures should be shown. The other way round (first crop&rotate then crop) is working as expected.

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version : 3.7.0+684~g40421d57d
* OS : Windows 10 // 19042.1165
* Memory : 16 GB
* Graphics card : Geforce RTX2070 + Intel UHD Graphics 630
* Graphics driver : 430.81
* OpenCL installed : yes
* OpenCL activated : yes

**Additional context**
_Please provide any additional information you think may be useful, for example:_

 - Can you reproduce with another darktable version(s)? **yes with version 3.6**
 - Are the steps above reproducible with a fresh edit (i.e. after discarding history)? **yes**

**Comment by johnny-bit** [comment]
This IS working as intended. Crop and rotate is before crop

**Comment by domosbg** [comment]
OK, understand. Because crop and rotate is in the pixel pipeline in front of crop, crop and rotate always works with the complete picture. I tried it again with this knowlegde and then it is reproducible for me what happens.

But if an unexperience user like me uses it the other way round the result is unexpected. Maybe it is possible to add a hint somewhere in the UI in crop and rotate if it is activated after crop module is already switched on like "Using 'crop and rotate' after 'crop' can lead to unexpected cropping areas".

**Comment by johnny-bit** [comment]
OK sounds ok... can you change the issue title and description to reflect that this is a Feature Request rather than bug report? And that you just want an alert about it? :)

**Comment by domosbg** [comment]
Changed the title and hope the new title is good.

What's the preffered way to show such an information/alert message? Should it be shown as an in picture overlay text, like the text that is shown when exporting pictures? 
![grafik](https://user-images.githubusercontent.com/88573684/130318364-00e5e626-3c27-4483-9aa2-0f76d09a7604.png)

I tried to take a look in the source code in order to understand how it works. If it is absolute nonsense what I figured out, then please ignore the following lines.
Are these messages showed by calling dt_control_log(...)?
The check if the alert message has to be shown has to be done in clipping.c in the function "void gui_focus(struct dt_iop_module_t *self, gboolean in)" in the part if (self->enabled) -> if(in) from line 1354?
Then unfortunately my understanding of the code ends. No idea, how to figure out if "crop" is already active.

**Comment by AlicVB** [comment]
well, the respective order of c&r and crop should NEVER be changed. That's a quite sensible area !
Ideally, we should not allow that at all, the same way, we restrict the move of some modules actually, but I'm not sure if it's possible to do that here... @TurboGit ??
That said, the idea is that sooner or later c&r will be deprecated in favor of crop and perspective modules. That still need some work to be done on some features, thought... see #9456 . so maybe we don't need to do too much work here...

**Comment by johnny-bit** [comment]
@AlicVB - the idea here is not much about moving modules in pipe but rather showing user warning if user wants to adjust crop&rotate after already adjusting crop module...

But I agree that functionality of crop&rotate might be better spent across crop & perspective modules. One additional problem then would be lightroom import which does use crop&rotate...

**Comment by TurboGit** [comment]
It is possible to fix some modules and ensure that they are not moved in the pipe. But I'm not sure there is circuitry to avoid swaping c&r and crop. Do do that we would need to add a specific rule, kind of a hack.

**Comment by AlicVB** [comment]
@TurboGit : thanks, I was right then...
Anyway, not worth the pain to change anything considering the state of the modules, imho

**Comment by domosbg** [comment]
@AlicVB @TurboGit I don't want to change anything in the pixel pipeline or fix modules. I changed the issue from a bug report to a feature request and because of that the issue might be be a little bit confusing now. Sorry for that. Next time I will try to make it more clearly. 

**I explain my suggestion again, then everthing is in one post:**
For a darktable beginner with no or not much knowledge of the exact pixel pipeline order, the result of two crops (first with the crop module and then with "crop and rotate") is normally not the result what these users are expecting.

So my suggestion is to show an alert like "Using 'crop and rotate' after 'crop' can lead to unexpected cropping results" message if someone focuses the "crop and rotate" module when the "crop" module is already switched on. Only as an information for the user that the result of what he is doing might not be what he expected.

I also tried to figure out how that alert could be implemented by looking at the source code. [My investigations are in this post](https://github.com/darktable-org/darktable/issues/9846#issuecomment-903094944). It is possible that it is complete nonsense. If so, please ignore it.

**Comment by AlicVB** [comment]
ah... yes, I understand now :) I've read your messages too quickly, sorry...
I'm not sure that what you suggest is possible, but I'll have a look

**Comment by domosbg** [comment]
@AlicVB I tried around a little bit. Maybe it would be enough to check if the crop module is in the history stack of the picture. I found such a piece of code in spot.c line 137 ff.

I tried to use that in clipping.c (sorry I'm too stupid to add correctly formatted code here, hope a screenshot also helps):
![grafik](https://user-images.githubusercontent.com/88573684/130369253-37da1df4-15c5-492c-b249-fb7f83b7b9ae.png)

Unfortunately if I try to debug it with a breakpoint, my eclipse IDE and darktable crashes after a few iterations of the for loop. If I let it run in debug mode without breakpoint, it seems to work in most cases. It also crashes one time, after that it worked. So I'm not sure if it is a problem with my eclipse installation or a problem with the code. I'm not familiar with C, so there might be problems with the usage of the pointers, that causes the problems.

Additionally to check if crop is in the history stack, a check if the crop module is also enabled would be nice, I think. There is an enabled variable in dt_dev_history_item_t which could be used for this check.

Sorry that I haven't finished my code snippet but I'm done for today. Installed the IDE and tried to understand a little bit of the code wasn't easy. Besides I think it is not a good idea if I made pull requests here with my knowledge of the C programming language. Won't like to damage something.
